### DISCUSSION TITLE
Any way to suppress tele/xxxx_06C6B8/RESULT MQTT messages

### DISCUSSION THREAD

**Discussion by UBWH** [arendst/Tasmota#20501] [Q&A]
I have a SCRIPT project that uses the `Smart Meter` feature to send/receive MODBUS messages. Works well.

Decided to port the project to BERRY.

Using `tasmota.cmd("modbussend {"deviceAddress" ...")` to send the MODBUS messages. Works well.
The responses come back in ModbusReceived messages as below.

```
11:45:28.551 MQT: tele/solar_06C6B8/RESULT = {"ModbusReceived":{"DeviceAddress":1,"FunctionCode":3,"StartAddress":36883,"Length":7,"Count":1,"Values":[2587]}}
11:45:29.552 MQT: tele/solar_06C6B8/RESULT = {"ModbusReceived":{"DeviceAddress":1,"FunctionCode":2,"StartAddress":8192,"Length":6,"Count":1,"Values":[41216]}}
11:45:30.552 MQT: tele/solar_06C6B8/RESULT = {"ModbusReceived":{"DeviceAddress":1,"FunctionCode":4,"StartAddress":13082,"Length":7,"Count":1,"Values":[1335]}}
11:45:31.557 MQT: tele/solar_06C6B8/RESULT = {"ModbusReceived":{"DeviceAddress":1,"FunctionCode":4,"StartAddress":13083,"Length":7,"Count":1,"Values":[-21]}}
11:45:32.510 MQT: tele/solar_06C6B8/RESULT = {"ModbusReceived":{"DeviceAddress":1,"FunctionCode":4,"StartAddress":12800,"Length":11,"Count":3,"Values":[0,7,1]}}
11:45:33.510 MQT: tele/solar_06C6B8/RESULT = {"ModbusReceived":{"DeviceAddress":1,"FunctionCode":4,"StartAddress":12560,"Length":9,"Count":2,"Values":[2937
```

As I am issuing, and receiving, MODBUS messages every second, I am getting MQTT messages emitted every second. 

I know these MQTT messages are not spam - that is how Tasmota works.

But - it is a major problem for me as these MQTT messages will be going over a low-bandwidth link to an external MQTT server that does not need them.

So my question is: How can I use BERRY to issue/receive MODBUS messages without so many MQTT messages?

(The SCRIPT method does not have this problem).

**Comment by sfromis** [comment]
No such option exist for the RESULT messages in general.

For serial bridge (not modbus), `SetOption147` exists to avoid `SerialReceived` messages being published via MQTT, but still available for local processing. It could make sense to extend this to where the Modbus bridge is calling `MqttPublishPrefixTopicRulesProcess_P` in [xdrv_63_modbus_bridge.ino](https://github.com/arendst/Tasmota/blob/development/tasmota/tasmota_xdrv_driver/xdrv_63_modbus_bridge.ino), in a similar way as what is being done in [xdrv_08_serial_bridge.ino](https://github.com/arendst/Tasmota/blob/development/tasmota/tasmota_xdrv_driver/xdrv_08_serial_bridge.ino#L177-L181)

**Reply by FulvioSpelta** [reply]
I think that your hint is very good.
Do you know if anyone has started to work on it ?

**Reply by sfromis** [reply]
I do not. And the step from mentioning something which would make sense to someone seeing it and decide to do something may not be "very direct" :wink:

**Reply by FulvioSpelta** [reply]
of course ! 😉

**Reply by FulvioSpelta** [reply]
Added the pull request #20678

**Reply by UBWH** [reply]
@FulvioSpelta 
Wow! Thanks for PR adding SetOption158 to suprress the MQTT messages on ModbusReceived messages

The way I have been capturing the Modbus replies (in Berry) is:
```
tasmota.add_rule("ModbusReceived",mbRX)    <--- mbRX() parses the ModbusReceived data
```

As a sanity check, if SetOption158 is set to ON, will my code above still work?

(I assume it will ... but thought I would double-check)

**Reply by FulvioSpelta** [reply]
Yes it should! Give it a try, if you can, and pls give back feedback.

**Reply by FulvioSpelta** [reply]
I've found a bug and submitted a PR to fix it #20733

**Reply by UBWH** [reply]
@FulvioSpelta 
I implemented the code like ...
```
if (Settings->flag6.mqtt_disable_modbus ) {  // SetOption158  If it is activated, Tasmota will not publish ModbusReceived MQTT messages, but it will proccess event trigger rules
            XdrvRulesProcess(0);
} else {
            MqttPublishPrefixTopicRulesProcess_P(RESULT_OR_TELE, PSTR(D_JSON_MODBUS_RECEIVED));
}
```

and it 100% supresses the `tele/solar_06C6B8/RESULT = {"ModbusReceived":{...` messages. 
Thanks.

But .. I am now getting lots of these MQTT messages (maybe always did ... but did not notice??)
```
stat/solar_06C6B8/RESULT = {"ModbusSend":"Done"}
```

The BERRY code the issues the commands is
```
tasmota.cmd("modbussend {"deviceAddress" ...")
```

Is there an easy way to supress these. (setoptionxxxx?)

**Reply by sfromis** [reply]
Suppressing everything should not be a goal, they do no harm. The point of avoiding the `ModbusReceived` messages being published is that these are intended as *the* way of processing the Modbus data, and when that processing is already done locally, that makes sense as a "takeover", in the same way as you can process a Switch or Button input with or without getting MQTT messages for these.

`ModbusSend` should be much lower in volume (indeed you always got them), and the command processing messages are inherent to how Tasmota commands are processed. Arguably, the message should've been more complete than just the `"Done"`, but that's a separate consideration.

`SetOption158` should already be part of the release code, thus you should not need to implement anything yourself.

**Reply by UBWH** [reply]
Appreciate all that, but as I said in the original post
`But - it is a major problem for me as these MQTT messages will be going over a low-bandwidth link to an external MQTT server ..`

**Reply by UBWH** [reply]
...See there was a new release 3 days ago (V14.4.0) with the new setoption158

**Reply by sfromis** [reply]
As I already told you
> No such option exist for the RESULT messages in general.

This is different from reporting of incoming Modbus messages, as the Tasmota *command* processing always elicits a response.

**Reply by FulvioSpelta** [reply]
I confirm that the setoption158 (pls take note that there is a fix in the development  branch #20733 ) suppress the "modbusreceived" messages. The "modbussend" messages follow a completely different rules and currently are not suppressed "by design".

I'll think a bit about this that really i don't know is this part of the tasmota code.
